### ISSUE TITLE
"Expected a null constant." error when saving assembly with pdb

### ISSUE THREAD

**Issue by maxkatz6** [0xd4d/dnlib#550]
I noticed this issue only happens when I used "local const" in the code, specifically when const type is an enum from another assembly. 
For example, this method:
```
public static bool IsMatch(string text, string pattern, CompareInfo compareInfo)
{
    const CompareOptions options = CompareOptions.IgnoreSymbols | CompareOptions.StringSort | CompareOptions.IgnoreNonSpace;
    return compareInfo.Compare(text, pattern, options) == 0;
}
```

Simply replacing `const CompareOptions options` with `CompareOptions options` will fix the problem, so we have a workaround, though this bug is confusing.
Repro project (both dll project and executable console app):

[ClassLibrary1.zip](https://github.com/0xd4d/dnlib/files/14739637/ClassLibrary1.zip)

**Comment by maxkatz6** [comment]
Also, unrelated to the original issue, and could be my misunderstanding, but I can't really run this code:
```
var filePath = args[0];
var asm = AssemblyDef.Load(filePath);
...
asm.Write(filePath, writerParameters);
```
as it will fail with `System.IO.IOException: The requested operation cannot be performed on a file with a user-mapped section open.`. Does dnLib hold file open after loading while trying to write to the same file?

**Comment by wtfsck** [comment]
It's memory mapped and you're using Windows which doesn't allow writing to the file until it's closed. You could instead read all bytes from the file and pass in the byte[] to the Load() method to load it that way and nothing gets memory mapped.

**Comment by wtfsck** [comment]
I think you need to add an assembly resolver to fix this. See the README on how to do that.

**Comment by maxkatz6** [comment]
> It's memory mapped and you're using Windows which doesn't allow writing to the file until it's closed. You could instead read all bytes from the file and pass in the byte[] to the Load() method to load it that way and nothing gets memory mapped.

Yeah, reading from byte array works fine. That's what I also did in the sample.

> I think you need to add an assembly resolver to fix this. See the README on how to do that.

Hm. In my actual project I did use an assembly resolver, where I initially found this issue. I trimmed everything down until I got this minimal sample.
I can try to update the repro of the issue, but it still fails with resolver. And removing `const` magically solves the issue.

**Comment by wtfsck** [comment]
Make sure the assembly resolver finds the assembly with the enum CompareOptions and any other assemblies it references, if any.

**Comment by UlyssesWu** [comment]
I met this issue, too. Could we just skip this error? Sometimes we cannot find every dll reference, but we still want some pdb info kept. However with this issue we can only get an empty pdb.

I have wrote some simple code to bypass this but I don't know if it will cause more problems, like making this pdb unusable.

https://github.com/0xd4d/dnlib/blob/1e0ec2607a146c17973e4e71051ff96d2ae7c2bc/src/DotNet/Pdb/Portable/LocalConstantSigBlobWriter.cs#L144-L152

```csharp
				case ElementType.Class when value is not byte[] && value is not null:
					var realElementType = ConstantUser.GetElementType(value);
					if (realElementType != ElementType.End) {
						writer.WriteByte((byte)realElementType);
						WritePrimitiveValue(writer, realElementType, value);
						WriteTypeDefOrRef(writer, type.ToTypeDefOrRef());
						return;
					}
					break;
```

**Comment by philipp-naused** [comment]
As a workaround, you can also remove the "invalid" constants from the module:

```cs
foreach (MethodDef method in new MemberFinder().FindAll(module).MethodDefs.Keys)
{
    IList<PdbConstant>? constants = method.Body?.PdbMethod?.Scope?.Constants;
    if (constants is not null)
    {
        for (int i = constants.Count - 1; i >= 0; i--)
        {
            // workaround for https://github.com/0xd4d/dnlib/issues/550
            // remove any constant with a reference type where the value is not a byte[] or null.
            PdbConstant constant = constants[i];
            if (constant.Type?.ElementType is ElementType.Class &&
                constant.Value is not null and not byte[])
            {
                constants.RemoveAt(i);
            }
        }
    }
}
```
